File: sdk/typescript/src/adp.ts
/**
 * ASD Distribution Protocol (ADP) — SAL-layer dictionary synchronization
 *
 * Complements the binary FNP handshake with SAL-level instructions for
 * version identity exchange, delta delivery, micro-delta, hash verification,
 * and MDR corpus version tracking.
 *
 * Patent ref: OSMP-001-UTIL Claims 20-21, Section VII.F, X-L
 * License: Apache 2.0
 */

import { AdaptiveSharedDictionary } from "./asd.js";
import { DictUpdateMode } from "./types.js";

// ── Version mapping: u16 wire as u8.u8 (MAJOR.MINOR) ──────────────────────

export function asdVersionPack(major: number, minor: number): number {
  if (major < 0 || major > 255 || minor < 0 || minor > 255)
    throw new RangeError(`Version ${major}.${minor} out of u8.u8 range`);
  return (major << 8) | minor;
}

export function asdVersionUnpack(u16: number): [number, number] {
  return [u16 >> 8, u16 & 0xFF];
}

export function asdVersionStr(u16: number): string {
  const [major, minor] = asdVersionUnpack(u16);
  return `${major}.${minor}`;
}

export function asdVersionParse(s: string): number {
  const parts = s.split(".");
  if (parts.length !== 2) throw new Error(`Invalid version string: ${s}`);
  return asdVersionPack(parseInt(parts[0], 10), parseInt(parts[1], 10));
}

export function asdVersionIsBreaking(oldU16: number, newU16: number): boolean {
  return (newU16 >> 8) > (oldU16 >> 8);
}

// ── Priority constants ─────────────────────────────────────────────────────

export const ADP_PRIORITY_MISSION  = 0;
export const ADP_PRIORITY_MICRO    = 1;
export const ADP_PRIORITY_DELTA    = 2;
export const ADP_PRIORITY_TRICKLE  = 3;

// ── Delta operation ────────────────────────────────────────────────────────

export interface ADPDeltaOp {
  namespace: string;
  mode: string;      // "+" | "\u2190" | "\u2020"
  opcode: string;
  definition?: string;
}

const MODE_NAMES: Record<string, string> = {
  "+": "ADDITIVE", "\u2190": "REPLACE", "\u2020": "DEPRECATE"
};

export function deltaOpToSal(op: ADPDeltaOp): string {
  return `${op.namespace}${op.mode}[${op.opcode}]`;
}

export function deltaOpIsBreaking(op: ADPDeltaOp): boolean {
  return op.mode === "\u2190";
}

// ── Delta payload ──────────────────────────────────────────────────────────

export interface ADPDelta {
  fromVersion: string;
  toVersion: string;
  operations: ADPDeltaOp[];
}

export function deltaToSal(d: ADPDelta): string {
  const ops = d.operations.map(deltaOpToSal).join(":");
  return `A:ASD:DELTA[${d.fromVersion}\u2192${d.toVersion}:${ops}]`;
}

export function deltaHasBreaking(d: ADPDelta): boolean {
  return d.operations.some(deltaOpIsBreaking);
}

// ── Pending instruction ────────────────────────────────────────────────────

export interface PendingInstruction {
  sal: string;
  unresolvedNamespace: string;
  unresolvedOpcode: string;
  timestamp: number;
}

// ── ADP Session ────────────────────────────────────────────────────────────

export class ADPSession {
  asd: AdaptiveSharedDictionary;
  asdVersion: number;
  namespaceVersions: Record<string, string>;
  pendingQueue: PendingInstruction[];
  deltaLog: string[];
  remoteVersion: number | null;
  remoteNamespaceVersions: Record<string, string> | null;

  constructor(
    asd: AdaptiveSharedDictionary,
    asdVersion: number = asdVersionPack(1, 0),
    namespaceVersions: Record<string, string> = {}
  ) {
    this.asd = asd;
    this.asdVersion = asdVersion;
    this.namespaceVersions = namespaceVersions;
    this.pendingQueue = [];
    this.deltaLog = [];
    this.remoteVersion = null;
    this.remoteNamespaceVersions = null;
  }

  // ── Version identity ───────────────────────────────────────────────

  versionIdentity(includeNamespaces: boolean = true): string {
    const ver = asdVersionStr(this.asdVersion);
    if (includeNamespaces && Object.keys(this.namespaceVersions).length > 0) {
      const ns = Object.keys(this.namespaceVersions).sort()
        .map(k => `:${k}${this.namespaceVersions[k]}`).join("");
      return `A:ASD[${ver}${ns}]`;
    }
    return `A:ASD[${ver}]`;
  }

  versionQuery(): string { return "A:ASD?"; }

  versionAlert(): string {
    return `A:ASD[${asdVersionStr(this.asdVersion)}]\u26a0`;
  }

  // ── Version parsing ────────────────────────────────────────────────

  receiveVersion(sal: string): {
    version: string; u16: number;
    namespaces: Record<string, string>;
    breaking: boolean; match: boolean;
  } {
    let inner = sal;
    if (inner.startsWith("A:ASD[")) inner = inner.slice(6);
    inner = inner.replace(/[\]\u26a0]+$/, "");

    const parts = inner.split(":");
    const verStr = parts[0];
    const remoteU16 = asdVersionParse(verStr);
    this.remoteVersion = remoteU16;

    const nsVersions: Record<string, string> = {};
    for (let i = 1; i < parts.length; i++) {
      const part = parts[i];
      if (part.length >= 2 && /[A-Z]/.test(part[0])) {
        nsVersions[part[0]] = part.slice(1);
      }
    }
    this.remoteNamespaceVersions = nsVersions;

    return {
      version: verStr,
      u16: remoteU16,
      namespaces: nsVersions,
      breaking: asdVersionIsBreaking(this.asdVersion, remoteU16),
      match: remoteU16 === this.asdVersion,
    };
  }

  // ── Delta request ──────────────────────────────────────────────────

  requestDelta(target?: string, namespace?: string): string {
    const myVer = asdVersionStr(this.asdVersion);
    const tgt = target || (this.remoteVersion !== null
      ? asdVersionStr(this.remoteVersion) : myVer);
    if (namespace && this.remoteNamespaceVersions) {
      const myNs = this.namespaceVersions[namespace] || "0.0";
      const remoteNs = this.remoteNamespaceVersions[namespace] || "0.0";
      return `A:ASD:REQ[${namespace}${myNs}\u2192${namespace}${remoteNs}]`;
    }
    return `A:ASD:REQ[${myVer}\u2192${tgt}]`;
  }

  // ── Delta application ──────────────────────────────────────────────

  applyDeltaSal(sal: string): {
    applied: boolean; from?: string; to?: string;
    operations?: string[]; breaking?: boolean;
    pendingResolved?: string[]; error?: string;
  } {
    this.deltaLog.push(sal);

    let inner = sal;
    if (inner.startsWith("A:ASD:DELTA[")) inner = inner.slice(12);
    inner = inner.replace(/\]+$/, "");

    const arrowIdx = inner.indexOf("\u2192");
    if (arrowIdx < 0) return { applied: false, error: "No version range" };

    const afterArrow = inner.slice(arrowIdx + 1);
    const colonIdx = afterArrow.indexOf(":");
    if (colonIdx < 0) return { applied: false, error: "No operations" };

    const fromVer = inner.slice(0, arrowIdx);
    const toVer = afterArrow.slice(0, colonIdx);
    const opsStr = afterArrow.slice(colonIdx + 1);

    const operations: string[] = [];
    let hasBreaking = false;
    const modeChars = new Set(["+", "\u2190", "\u2020"]);

    let pos = 0;
    while (pos < opsStr.length) {
      if (!/[A-Z]/.test(opsStr[pos])) { pos++; continue; }
      const ns = opsStr[pos]; pos++;
      if (pos >= opsStr.length) break;
      const mode = opsStr[pos];
      if (!modeChars.has(mode)) continue;
      pos++;
      if (pos < opsStr.length && opsStr[pos] === "[") {
        const endBracket = opsStr.indexOf("]", pos);
        const opcode = endBracket >= 0
          ? opsStr.slice(pos + 1, endBracket)
          : opsStr.slice(pos + 1);
        pos = endBracket >= 0 ? endBracket + 1 : opsStr.length;

        const asdMode = mode === "+"
          ? DictUpdateMode.ADDITIVE
          : mode === "\u2190"
            ? DictUpdateMode.REPLACE
            : DictUpdateMode.DEPRECATE;
        this.asd.applyDelta(ns, opcode, "", asdMode, toVer);
        if (mode === "\u2190") hasBreaking = true;
        operations.push(`${ns}:${opcode}(${MODE_NAMES[mode]})`);
      }
    }

    const resolved = this.resolvePending();

    return {
      applied: true, from: fromVer, to: toVer,
      operations, breaking: hasBreaking,
      pendingResolved: resolved,
    };
  }

  // ── Micro-delta ────────────────────────────────────────────────────

  requestDefinition(namespace: string, opcode: string): string {
    return `A:ASD:DEF?[${namespace}:${opcode}]`;
  }

  sendDefinition(namespace: string, opcode: string,
                  definition: string, layer: number = 1): string {
    return `A:ASD:DEF[${namespace}:${opcode}:${definition}:${layer}]`;
  }

  applyDefinition(sal: string): {
    applied: boolean; namespace?: string; opcode?: string;
    definition?: string; layer?: number;
    pendingResolved?: string[]; error?: string;
  } {
    this.deltaLog.push(sal);
    let inner = sal;
    if (inner.startsWith("A:ASD:DEF[")) inner = inner.slice(10);
    inner = inner.replace(/\]+$/, "");

    const parts = inner.split(":");
    if (parts.length < 3) return { applied: false, error: "Insufficient fields" };

    const [namespace, opcode, definition] = parts;
    const layer = parts.length > 3 ? parseInt(parts[3], 10) : 1;

    this.asd.applyDelta(namespace, opcode, definition,
                         DictUpdateMode.ADDITIVE, "micro");

    const resolved = this.resolvePending();

    return {
      applied: true, namespace, opcode, definition, layer,
      pendingResolved: resolved,
    };
  }

  // ── Hash verification ──────────────────────────────────────────────

  hashIdentity(hexLength: number = 8): string {
    const ver = asdVersionStr(this.asdVersion);
    const fp = this.asd.fingerprint().slice(0, hexLength);
    return `A:ASD:HASH[${ver}:${fp}]`;
  }

  verifyHash(sal: string): { match: boolean; remoteHash?: string; localHash?: string } {
    let inner = sal;
    if (inner.startsWith("A:ASD:HASH[")) inner = inner.slice(11);
    inner = inner.replace(/\]+$/, "");
    const parts = inner.split(":");
    if (parts.length < 2) return { match: false };
    const remoteHash = parts[1];
    const localHash = this.asd.fingerprint().slice(0, remoteHash.length);
    return { match: remoteHash === localHash, remoteHash, localHash };
  }

  // ── MDR corpus versioning ──────────────────────────────────────────

  static mdrIdentity(corpora: Record<string, string>): string {
    const parts = Object.keys(corpora).sort()
      .map(k => `${k}:${corpora[k]}`).join(":");
    return `A:MDR[${parts}]`;
  }

  static mdrRequest(corpus: string, fromVer: string, toVer: string): string {
    return `A:MDR:REQ[${corpus}:${fromVer}\u2192${toVer}]`;
  }

  // ── Semantic pending queue ─────────────────────────────────────────

  resolveOrPend(sal: string): {
    resolved: boolean; pending: boolean;
    definition?: string; unresolved?: string;
    microDeltaRequest?: string; queueDepth?: number;
  } {
    const [ns, opcode] = ADPSession.extractNsOpcode(sal);
    if (ns === null) return { resolved: true, pending: false };

    const definition = this.asd.lookup(ns, opcode!);
    if (definition !== null)
      return { resolved: true, pending: false, definition };

    this.pendingQueue.push({
      sal, unresolvedNamespace: ns,
      unresolvedOpcode: opcode!,
      timestamp: Date.now(),
    });

    return {
      resolved: false, pending: true,
      unresolved: `${ns}:${opcode}`,
      microDeltaRequest: this.requestDefinition(ns, opcode!),
      queueDepth: this.pendingQueue.length,
    };
  }

  private resolvePending(): string[] {
    const resolved: string[] = [];
    const stillPending: PendingInstruction[] = [];
    for (const p of this.pendingQueue) {
      if (this.asd.lookup(p.unresolvedNamespace, p.unresolvedOpcode) !== null)
        resolved.push(p.sal);
      else
        stillPending.push(p);
    }
    this.pendingQueue = stillPending;
    return resolved;
  }

  private static extractNsOpcode(sal: string): [string | null, string | null] {
    if (!sal || !/^[A-Z]/.test(sal) || !sal.includes(":"))
      return [null, null];
    const parts = sal.split(":");
    if (parts.length < 2 || parts[0].length !== 1) return [null, null];
    let opcode = "";
    for (const ch of parts[1]) {
      if ("[]?<>@\u2227\u2228\u2192\u26a0".includes(ch)) break;
      opcode += ch;
    }
    return [parts[0], opcode || null];
  }

  // ── Acknowledge ────────────────────────────────────────────────────

  static acknowledgeVersion(version: string): string {
    return `A:ACK[ASD:${version}]`;
  }
  static acknowledgeHash(): string { return "A:ACK[ASD:HASH]"; }
  static acknowledgeDef(): string { return "A:ACK[ASD:DEF]"; }

  // ── Priority classification ────────────────────────────────────────

  static classifyPriority(sal: string): number {
    if (!sal.startsWith("A:ASD") && !sal.startsWith("A:MDR"))
      return ADP_PRIORITY_MISSION;
    if (sal.includes("DEF")) return ADP_PRIORITY_MICRO;
    if (sal.includes("DELTA")) return ADP_PRIORITY_DELTA;
    return ADP_PRIORITY_TRICKLE;
  }
}

File: sdk/go/osmp/adp.go
package osmp

// ASD Distribution Protocol (ADP) — SAL-layer dictionary synchronization
//
// Complements the binary FNP handshake with SAL-level instructions for
// version identity exchange, delta delivery, micro-delta, hash verification,
// and MDR corpus version tracking.
//
// Patent ref: OSMP-001-UTIL Claims 20-21, Section VII.F, X-L
// License: Apache 2.0

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
	"time"
)

// ── Version mapping: u16 wire as u8.u8 (MAJOR.MINOR) ──────────────────────

// ASDVersionPack encodes MAJOR.MINOR into u16 for FNP wire format.
func ASDVersionPack(major, minor int) (uint16, error) {
	if major < 0 || major > 255 || minor < 0 || minor > 255 {
		return 0, fmt.Errorf("version %d.%d out of u8.u8 range", major, minor)
	}
	return uint16(major<<8 | minor), nil
}

// ASDVersionUnpack decodes u16 from FNP wire format into (major, minor).
func ASDVersionUnpack(u16 uint16) (int, int) {
	return int(u16 >> 8), int(u16 & 0xFF)
}

// ASDVersionStr returns display string for SAL instructions: "2.7".
func ASDVersionStr(u16 uint16) string {
	major, minor := ASDVersionUnpack(u16)
	return fmt.Sprintf("%d.%d", major, minor)
}

// ASDVersionParse parses "2.7" into u16.
func ASDVersionParse(s string) (uint16, error) {
	parts := strings.Split(s, ".")
	if len(parts) != 2 {
		return 0, fmt.Errorf("invalid version string: %s", s)
	}
	major, err1 := strconv.Atoi(parts[0])
	minor, err2 := strconv.Atoi(parts[1])
	if err1 != nil || err2 != nil {
		return 0, fmt.Errorf("invalid version string: %s", s)
	}
	return ASDVersionPack(major, minor)
}

// ASDVersionIsBreaking returns true if the version change includes a MAJOR increment.
func ASDVersionIsBreaking(oldU16, newU16 uint16) bool {
	return (newU16 >> 8) > (oldU16 >> 8)
}

// ── Priority constants ─────────────────────────────────────────────────────

const (
	ADPPriorityMission = 0 // Non-ADP instruction traffic
	ADPPriorityMicro   = 1 // Task-relevant micro-delta (A:ASD:DEF)
	ADPPriorityDelta   = 2 // Background delta (A:ASD:DELTA)
	ADPPriorityTrickle = 3 // Trickle charge request (A:ASD:REQ, A:ASD)
)

// ── Delta operation ────────────────────────────────────────────────────────

// ADPDeltaOp represents a single operation within a delta payload.
type ADPDeltaOp struct {
	Namespace  string
	Mode       string // "+" | "←" | "†"
	Opcode     string
	Definition string
}

// IsBreaking returns true if this is a REPLACE operation.
func (op *ADPDeltaOp) IsBreaking() bool { return op.Mode == "\u2190" }

// ToSAL returns the SAL representation: "H+[LACTATE]".
func (op *ADPDeltaOp) ToSAL() string {
	return fmt.Sprintf("%s%s[%s]", op.Namespace, op.Mode, op.Opcode)
}

// ── Delta payload ──────────────────────────────────────────────────────────

// ADPDelta represents a complete delta payload with version range and operations.
type ADPDelta struct {
	FromVersion string
	ToVersion   string
	Operations  []ADPDeltaOp
}

// HasBreaking returns true if any operation is a REPLACE.
func (d *ADPDelta) HasBreaking() bool {
	for _, op := range d.Operations {
		if op.IsBreaking() {
			return true
		}
	}
	return false
}

// ToSAL returns the full SAL instruction.
func (d *ADPDelta) ToSAL() string {
	ops := make([]string, len(d.Operations))
	for i, op := range d.Operations {
		ops[i] = op.ToSAL()
	}
	return fmt.Sprintf("A:ASD:DELTA[%s\u2192%s:%s]",
		d.FromVersion, d.ToVersion, strings.Join(ops, ":"))
}

// ── Pending instruction ────────────────────────────────────────────────────

// PendingInstruction is held in the semantic pending queue.
type PendingInstruction struct {
	SAL                 string
	UnresolvedNamespace string
	UnresolvedOpcode    string
	Timestamp           time.Time
}

// ── ADP Session ────────────────────────────────────────────────────────────

// ADPSession manages SAL-layer dictionary synchronization.
type ADPSession struct {
	ASD                    *AdaptiveSharedDictionary
	ASDVersion             uint16
	NamespaceVersions      map[string]string
	PendingQueue           []PendingInstruction
	DeltaLog               []string
	RemoteVersion          *uint16
	RemoteNamespaceVersions map[string]string
}

// NewADPSession creates a new ADP session manager.
func NewADPSession(asd *AdaptiveSharedDictionary, asdVersion uint16,
	nsVersions map[string]string) *ADPSession {
	if nsVersions == nil {
		nsVersions = make(map[string]string)
	}
	return &ADPSession{
		ASD:               asd,
		ASDVersion:        asdVersion,
		NamespaceVersions: nsVersions,
		PendingQueue:      nil,
		DeltaLog:          nil,
	}
}

// VersionIdentity generates A:ASD version identity instruction.
func (s *ADPSession) VersionIdentity(includeNamespaces bool) string {
	ver := ASDVersionStr(s.ASDVersion)
	if includeNamespaces && len(s.NamespaceVersions) > 0 {
		keys := make([]string, 0, len(s.NamespaceVersions))
		for k := range s.NamespaceVersions {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		var ns strings.Builder
		for _, k := range keys {
			ns.WriteString(fmt.Sprintf(":%s%s", k, s.NamespaceVersions[k]))
		}
		return fmt.Sprintf("A:ASD[%s%s]", ver, ns.String())
	}
	return fmt.Sprintf("A:ASD[%s]", ver)
}

// VersionQuery generates version query broadcast.
func (s *ADPSession) VersionQuery() string { return "A:ASD?" }

// VersionAlert generates version update announcement.
func (s *ADPSession) VersionAlert() string {
	return fmt.Sprintf("A:ASD[%s]\u26a0", ASDVersionStr(s.ASDVersion))
}

// RequestDelta generates a delta request instruction.
func (s *ADPSession) RequestDelta(target string) string {
	myVer := ASDVersionStr(s.ASDVersion)
	return fmt.Sprintf("A:ASD:REQ[%s\u2192%s]", myVer, target)
}

// RequestDefinition generates a micro-delta request.
func (s *ADPSession) RequestDefinition(namespace, opcode string) string {
	return fmt.Sprintf("A:ASD:DEF?[%s:%s]", namespace, opcode)
}

// SendDefinition generates a micro-delta response.
func (s *ADPSession) SendDefinition(namespace, opcode, definition string, layer int) string {
	return fmt.Sprintf("A:ASD:DEF[%s:%s:%s:%d]", namespace, opcode, definition, layer)
}

// HashIdentity generates hash verification instruction.
func (s *ADPSession) HashIdentity(hexLength int) string {
	ver := ASDVersionStr(s.ASDVersion)
	fp := s.ASD.Fingerprint()
	if len(fp) > hexLength {
		fp = fp[:hexLength]
	}
	return fmt.Sprintf("A:ASD:HASH[%s:%s]", ver, fp)
}

// ResolveOrPend checks if an instruction's opcodes are resolvable.
// If not, adds it to the semantic pending queue.
func (s *ADPSession) ResolveOrPend(sal string) (resolved bool, pending bool, microReq string) {
	ns, opcode := extractNsOpcode(sal)
	if ns == "" {
		return true, false, ""
	}
	def := s.ASD.Lookup(ns, opcode)
	if def != "" {
		return true, false, ""
	}
	s.PendingQueue = append(s.PendingQueue, PendingInstruction{
		SAL: sal, UnresolvedNamespace: ns,
		UnresolvedOpcode: opcode, Timestamp: time.Now(),
	})
	return false, true, s.RequestDefinition(ns, opcode)
}

func extractNsOpcode(sal string) (string, string) {
	if len(sal) == 0 || sal[0] < 'A' || sal[0] > 'Z' || !strings.Contains(sal, ":") {
		return "", ""
	}
	parts := strings.SplitN(sal, ":", 3)
	if len(parts) < 2 || len(parts[0]) != 1 {
		return "", ""
	}
	opRaw := parts[1]
	opcode := ""
	for _, ch := range opRaw {
		if strings.ContainsRune("[]?<>@\u2227\u2228\u2192\u26a0", ch) {
			break
		}
		opcode += string(ch)
	}
	return parts[0], opcode
}

// ClassifyPriority returns the ADP priority level for a SAL instruction.
func ClassifyPriority(sal string) int {
	if !strings.HasPrefix(sal, "A:ASD") && !strings.HasPrefix(sal, "A:MDR") {
		return ADPPriorityMission
	}
	if strings.Contains(sal, "DEF") {
		return ADPPriorityMicro
	}
	if strings.Contains(sal, "DELTA") {
		return ADPPriorityDelta
	}
	return ADPPriorityTrickle
}

// MDR helpers

// MDRIdentity generates MDR corpus version identity.
func MDRIdentity(corpora map[string]string) string {
	keys := make([]string, 0, len(corpora))
	for k := range corpora {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	parts := make([]string, len(keys))
	for i, k := range keys {
		parts[i] = fmt.Sprintf("%s:%s", k, corpora[k])
	}
	return fmt.Sprintf("A:MDR[%s]", strings.Join(parts, ":"))
}

// MDRRequest generates MDR delta request.
func MDRRequest(corpus, fromVer, toVer string) string {
	return fmt.Sprintf("A:MDR:REQ[%s:%s\u2192%s]", corpus, fromVer, toVer)
}

// Acknowledge helpers

func AcknowledgeVersion(version string) string { return fmt.Sprintf("A:ACK[ASD:%s]", version) }
func AcknowledgeHash() string                  { return "A:ACK[ASD:HASH]" }
func AcknowledgeDef() string                   { return "A:ACK[ASD:DEF]" }
